bundles/cif-connector-graphql/src/main/java/com/adobe/cq/commerce/graphql/magento/GraphqlDataServiceImpl.java

```diff
@@ -74,31 +74,48 @@
     private Cache<String, Optional<ProductInterface>> productCache;
     private Cache<String, Optional<CategoryProducts>> categoryCache;
 
-    protected Map<String, GraphqlClient> clients = new ConcurrentHashMap<>();
+    private Map<String, GraphqlClient> clients = new ConcurrentHashMap<>();
 
     @Reference(
         service = GraphqlClient.class,
         bind = "bindGraphqlClient",
         unbind = "unbindGraphqlClient",
-        cardinality = ReferenceCardinality.AT_LEAST_ONE,
+        cardinality = ReferenceCardinality.MULTIPLE,
         policy = ReferencePolicy.DYNAMIC)
     protected void bindGraphqlClient(GraphqlClient graphqlClient, Map<?, ?> properties) {
         String identifier = graphqlClient.getIdentifier();
         LOGGER.info("Registering GraphqlClient '{}'", identifier);
         clients.put(identifier, graphqlClient);
+
+        if (configuration.identifier().equals(identifier)) {
+            LOGGER.info("GraphqlClient with identifier '{}' has been registered, the service is ready to handle requests.", identifier);
+            baseClient = graphqlClient;
+        }
     }
 
     protected void unbindGraphqlClient(GraphqlClient graphqlClient, Map<?, ?> properties) {
         String identifier = graphqlClient.getIdentifier();
         LOGGER.info("De-registering GraphqlClient '{}'", identifier);
         clients.remove(identifier);
+
+        if (configuration.identifier().equals(identifier)) {
+            LOGGER.info("GraphqlClient '{}' unregistered: requests cannot be handled until that dependency is satisfied", identifier);
+            baseClient = null;
+        }
     }
 
     @Activate
     public void activate(GraphqlDataServiceConfiguration conf) throws Exception {
         baseClient = clients.get(conf.identifier());
         if (baseClient == null) {
-            throw new RuntimeException("Cannot find GraphqlClient with identifier '" + conf.identifier() + "'");
+            // Because of the many:many OSGi dynamic dependencies between GraphqlDataServiceImpl and GraphqlClientImpl, we cannot enforce
+            // the dependency and hence services might not start in the right order. So we ignore the missing client, it might start later
+            // and would be set in the bindGraphqlClient() method.
+            // Important: we let all public methods fail with NullPointerException, so that the Sling ResourceProvider will get exceptions
+            // when trying to get resources. This is important, because it ensures that the resource provider will always try to refetch
+            // the resources (if we for example return an empty list, this would get cached by Sling).
+
+            LOGGER.warn("GraphqlClient '{}' not found: requests cannot be handled until that dependency is satisfied", conf.identifier());
```

Comment:
I suggest this should be an error log as something is not working.

Reply:
Well it's not really an error because it can recover from it as soon as the graphql client is started, so it's rather an "expected temporary issue".

Reply:
Ok then makes sense.